core/src/main/scala/kafka/server/KafkaApis.scala

```
@@ -2468,7 +2483,7 @@ class KafkaApis(val requestChannel: RequestChannel,
           } else {
             // Otherwise, the regular appendRecords path is used for all the non __consumer_offsets
             // partitions or for all partitions when the new group coordinator is disabled.
-            controlRecords += partition -> MemoryRecords.withEndTransactionMarker(
+            controlRecords += replicaManager.topicIdPartition(partition) -> MemoryRecords.withEndTransactionMarker(
```

`replicaManager.topicIdPartition` could have zero for topicId. Should we return an error in that case? Every caller to `replicaManager.appendRecords` needs to deal with this issue. Is it better to deal with this inside `replicaManager.appendRecords` or in every caller?

---

I think `appendRecords` should handle so we handle this in one place. I'll update `appendRecords` to reflect this

---

Thought about this and I am not clear why the caller needs to deal with this unless I am missing something here? if topicId is set to zero we always will fall back to the old behaviour as if this is an old request.

---

We could leave it as it is. However, it could be useful to document for `replicaManager.appendRecords` whether the passed in partitions can have topicId with 0 or not.

---

I have updated `replicaManager.appendRecords` java doc and left a comment here as well

---

For internal appends, we should always be able to use topic ID though right? It's ok to fall back to topic name behavior, but just wanted to point this out. (In theory I guess the WriteTxnMarkers api could pass topic IDs, but then we would have to store topic IDs in the txn manager 😁 )

---

This is actually mostly for a user topic. But it's possible for the user or the internal topic to be deleted and we need a path to handle that rare case.

---

Sorry was not clear for "internal appends" as I meant writing to a partition (user or not) not via the produce api, but just append records from within the broker. The point being, we will just grab what the broker sees as the topic id and not the "original" topic ID added to the transaction.

---

Yes, if AddPartitionsToTxnRequest and WriteTxnMarkersRequest both use topicId, we could use it directly here.

Do you still have any concerns on this change?

---

No concerns, just that this doesn't do this checking...yet :) 

If we want to just pull the topic ID from the ReplicaManager for now that's fine.

---

What checking are you referring to?

---

Checking that the topic ID is the same. As a followup we could do that.